onnxscript/ir/_core.py

```diff
@@ -1390,19 +1427,46 @@ def extend(self, nodes: Iterable[Node], /) -> None:
         nodes = [self._set_node_graph_to_self_and_assign_names(node) for node in nodes]
         self._nodes.extend(nodes)
 
-    def remove(self, node: Node, /) -> None:
-        """Remove a node from the graph in O(1) time.
+    def remove(self, nodes: Node | Iterable[Node], /, safe: bool = False) -> None:
+        """Remove nodes from the graph in O(#num of nodes) time.
+
+        If any errors are raise, to ensure the graph is not left in an inconsistent state,
+        the graph is not modified.
 
         Args:
-            node: The node to remove.
+            nodes: The node to remove.
+            safe: If True, performs the following actions before removal:
+                1. It checks to make sure there are no users of the node that are not
+                    to be removed before removing it.
+                2. It checks the node does not contribute to any graph outputs.
+                3. It removes references to all inputs so it is no longer a user of other nodes.
+
+                This option will break all interconnections (input/output) of the node to remove.
 
         Raises:
-            ValueError: If the node does not belong to this graph.
+            ValueError: If any node to remove does not belong to this graph.
+            ValueError: (When ``safe=True``) If the node does not belong to this graph or if there are users of the node.
+            ValueError: (When ``safe=True``) If the node is still being used by other nodes not to be removed.
         """
-        if node.graph is not self:
-            raise ValueError(f"The node {node} does not belong to this graph.")
-        node.graph = None
-        self._nodes.remove(node)
+        if not isinstance(nodes, Iterable):
+            nodes = {nodes}
+        else:
+            nodes = frozenset(nodes)
+        graph_outputs = frozenset(self.outputs)
+        for node in nodes:
+            if node.graph is not self:
+                raise ValueError(f"The node {node} does not belong to this graph.")
+            if safe:
+                # Check 1, 2
+                _check_node_safe_to_remove(node, nodes, graph_outputs)
```

Comment:
``nodes`` can be an iterator here and ``_check_node_safe_to_remove`` checks that the consumers of `node` belongs to ``nodes``. Maybe Iterator is not the right annotation?

Reply:
I think it is made a set in https://github.com/microsoft/onnxscript/pull/1415/files#diff-33a5ea7182ac2a3ba4127e9fea3cd2db2dbe54810994ca210727d28fccf2b5cfR1451-R1455. Do you prefer the new variable to be renamed for clarity?

Reply:
Renamed to nodes_set